File: SwiftDemangler/SwiftDemangler.java
// This script demangle swift function names using ghidra2frida bridge, 
// running the demangling code directly on the mobile iOS device. The  
// Java code is based on Ghidra Ninja swift_demangler.py 
// (https://github.com/ghidraninja/ghidra_scripts/blob/master/swift_demangler.py)
// @author Federico Dotta
// @category iOS
// @keybinding
// @menupath Tools.SwiftDemangler
// @toolbar 


/*
 * SwiftDemangler - Demangle Swift functions with ghidra2frida (with service import)
 *
 * Tested with Ghidra v10.0.1
 */


import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.FunctionManager;

import ghidra2frida.Ghidra2FridaService;

public class SwiftDemangler extends GhidraScript {
	
	Ghidra2FridaService ghidra2FridaService;
	
	boolean debug=false;
	
	public String callGhidra2FridaFunction(String name, String[] parameters) throws Exception {
		
		if(ghidra2FridaService == null)
			ghidra2FridaService = state.getTool().getService(Ghidra2FridaService.class);
		
		return ghidra2FridaService.callExportedFunction("demangle",parameters);		
		
	}
	
	@Override
	protected void run() throws Exception {
		
		FunctionManager functionManager = currentProgram.getFunctionManager();
		
		for(FunctionIterator fns = functionManager.getFunctions(true); fns.hasNext(); ) {
			
			Function f = fns.next();
			
			String f_name = f.getName();
			
			//println(f_name);
			
			// Is it a mangled name?
			if(!(f_name.startsWith("_$") || f_name.startsWith("$s") || f_name.startsWith("_T") || f_name.startsWith("__T")))
				continue;
			
			if(f_name.startsWith("__T"))
				f_name = f_name.substring(1);
			
			
			String previous_comment = f.getComment();
		    if(previous_comment == null)
		        previous_comment = "";

		    try {
		    			    
			    String signature_full = callGhidra2FridaFunction("demangle",new String[] {f_name});
			    
			    String signature = signature_full;
			    
			    if(debug)
			    	println("Mangled: " + f_name + ", Demangled: " + signature_full);
			    			    
			    // Add newlines into full comment (maximum comment len = 58, afterwards truncated)
			    int lines = signature_full.length() / 58;
			    for(int l=1; l < lines+1; l++) 
			        signature_full = signature_full.substring(0, (l*58)+(l-1)) + "\n" + signature_full.substring((l*58)+(l-1));
	
			    f.setComment(previous_comment + "\n" + signature_full);
			    
			    
		    } catch(Exception e) {
		    	
		    	printerr("* Error with method: " + f_name + ". Skipping. Exception:");
		    	printerr(e.toString());
		    	continue;
		    	
		    }
		    
			
		}
		
	}
}

File: SwiftDemangler/swift_demangler.py
#This script demangle swift function names using ghidra2frida bridge, 
#running the demangling code directly on the mobile iOS device. The  
#code is based on Ghidra Ninja swift_demangler.py 
#(https://github.com/ghidraninja/ghidra_scripts/blob/master/swift_demangler.py)
#@author Federico Dotta
#@category iOS
#@keybinding
#@menupath Tools.SwiftDemangler
#@toolbar 

from ghidra.program.model.symbol import SourceType

from ghidra2frida import Ghidra2FridaService

debug_mode = False

ghidra2FridaService = None

def callGhidra2FridaFunction(name, parameters):
    global ghidra2FridaService
    if ghidra2FridaService is None:
        ghidra2FridaService = state.getTool().getService(Ghidra2FridaService)
        println("ghidra2FridaService initialized")

    return ghidra2FridaService.callExportedFunction(name,parameters);

println("ghidra2frida demangle START")

functionManager = currentProgram.getFunctionManager()

# Get functions in ascending order
fns = functionManager.getFunctions(True)
for f in fns:
    f_name = f.getName()

    if debug_mode:
        println("Processing " + f_name)

    # Is it a mangled name?
    if not (f_name.startswith("_$") or f_name.startswith("$s") or f_name.startswith("_T") or f_name.startswith("__T")):
        continue

    if f_name.startswith("__T"):
        f_name =  f_name[1:]        

    previous_comment = f.getComment()

    try:

        signature_full = callGhidra2FridaFunction("demangle",[f_name]);

        if debug_mode:
            println("Demangle output: " + signature_full)

        # Add newlines into full comment (maximum comment len = 58, afterwards truncated)
        lines = len(signature_full) / 58
        for l in range(1, lines+1):
            signature_full = signature_full[:(l*58)+(l-1)] + "\n" + signature_full[(l*58)+(l-1):]

        if not previous_comment:
            f.setComment(signature_full)
        else:    
            f.setComment(previous_comment + "\n" + signature_full)
        
    except Exception as exc:
        printerr("Error with method: " + f_name + ". Skipping. Exception:")    
        printerr(str(exc))


println("ghidra2frida demangle END")